src/number_theory/bertrand.lean

```diff
@@ -0,0 +1,251 @@
+end
+
+end nat
+
+end bertrand
+
+/--
+Proves that Bertrand's postulate holds for all sufficiently large `n`.
+-/
+lemma bertrand_eventually (n : ℕ) (n_big : 512 ≤ n) : ∃ (p : ℕ), p.prime ∧ n < p ∧ p ≤ 2 * n :=
+begin
+  -- Assume there is no prime in the range.
+  by_contradiction no_prime,
+  -- Then we have the above sub-exponential bound on the size of this central binomial coefficient.
+  -- We now couple this bound with an exponential lower bound on the central binomial coefficient,
+  -- yielding an inequality which we have seen is false for large enough n.
+  linarith [bertrand.main_inequality n_big, nat.four_pow_lt_mul_central_binom n (by linarith),
+    mul_le_mul_left' (bertrand.central_binom_le_of_no_bertrand_prime n (by linarith) no_prime) n],
+end
+
+/--
+Proves that Bertrand's postulate holds over all positive naturals less than n by identifying a
+descending list of primes, each no more than twice the next, such that the list contains a witness
+for each number ≤ n.
+-/
+lemma bertrand_initial (n : ℕ) (hn0 : 0 < n)
+  (plist : list ℕ) (prime_plist : ∀ p ∈ plist, nat.prime p)
+  (covering : ∀ a b, (a, b) ∈ list.zip plist (list.tail (plist ++ [2])) → a ≤ 2 * b)
+  (hn : n < (plist ++ [2]).head) :
+  ∃ (p : ℕ), p.prime ∧ n < p ∧ p ≤ 2 * n :=
+begin
+  unfreezingI { induction plist, },
+  { simp only [list.nil_append, list.head_cons] at hn,
+    interval_cases n,
+    { use 2,
+      norm_num, }, },
+  { simp only [list.mem_cons_iff, forall_eq_or_imp] at prime_plist,
+    by_cases plist_hd ≤ 2 * n,
+    { use plist_hd,
+      exact ⟨prime_plist.left, hn, h⟩, },
+    { apply plist_ih prime_plist.right,
+      { intros a b hmem,
+        apply covering,
+        cases plist_tl,
+        { simp only [list.nil_append, list.tail_cons, list.zip_nil_right, list.not_mem_nil] at hmem,
+          exfalso, exact hmem, },
+        { simp only [list.cons_append, list.tail_cons] at hmem,
+          right,
+          exact hmem, }, },
+      { cases plist_tl,
+        { have h_hd := covering plist_hd 2,
+          simp only [list.singleton_append, list.tail_cons, list.zip_cons_cons, list.zip_nil_right,
+            list.mem_singleton, eq_self_iff_true, forall_true_left] at h_hd,
+          have hn2 : 2 * n < 2 * 2 := by linarith only [h, h_hd],
+          exact lt_of_mul_lt_mul_left' hn2, },
+        { simp only [list.cons_append, list.head_cons],
+          have h_hd := covering plist_hd plist_tl_hd,
+          simp only [list.cons_append, list.tail_cons, list.zip_cons_cons, list.mem_cons_iff,
+            eq_self_iff_true, true_or, forall_true_left] at h_hd,
+          rw [not_le] at h,
+          have hn2 : 2 * n < 2 * plist_tl_hd, exact gt_of_ge_of_gt h_hd h,
+          exact lt_of_mul_lt_mul_left' hn2, }, }, }, },
+end
+
+/--
+Bertrand's Postulate: For any positive natural number, there is a prime which is greater than
+it, but no more than twice as large.
+-/
+theorem bertrand (n : ℕ) (n_pos : 0 < n) : ∃ p, nat.prime p ∧ n < p ∧ p ≤ 2 * n :=
```

Comment:
We usually assume `n ≠ 0` because it looks like a weaker assumption.

Reply:
Should it be called `nat.exists_prime_lt_and_le_two_mul` (with an alias `nat.bertrand`)?

Reply:
changed